### PR TITLE
Adds tests for cardinality and filter aggregations

### PR DESCRIPTION
**Pull request by colings86**
Relates to #22278

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java [hunk 1]
```diff
@@ -420,6 +423,33 @@ public void close() {
...
+        Set<Object> values = new HashSet<>();
+        if (algorithm.get(bucket) == LINEAR_COUNTING) {
+            try (IntArray hashSetValues = hashSet.values(bucket)) {
+                for (long i = 0; i < hashSetValues.size(); i++) {
+                    values.add(hashSetValues.get(i));
+                }
+            }
+        } else {
+            for (long i = 0; i < runLens.size(); i++) {
+                values.add(runLens.get(i));
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java]: I think that should be s/i/(bucket << p)+i/

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java [hunk 2]
```diff
@@ -420,6 +423,33 @@ public void close() {
...
+    }
+
+    public int hashCode(long bucket) {
+        return Objects.hash(p, algorithm.get(bucket), getComparableData(bucket));
+    }
+
+    public boolean equals(long bucket, HyperLogLogPlusPlus other) {
+        return Objects.equals(p, other.p) &&
+                Objects.equals(algorithm.get(bucket), other.algorithm.get(bucket))
+                &&
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java]: weird indentation?

File: core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java [hunk 3]
```diff
@@ -578,6 +608,23 @@ void clear(long bit) {
...
+        public boolean equals(Object obj) {
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            OpenBitSet other = (OpenBitSet) obj;
+            return Objects.equals(impl, other.impl);
+        }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/aggregations/metrics/cardinality/HyperLogLogPlusPlus.java]: It seems to me that this is not needed anymore?

File: core/src/test/java/org/elasticsearch/search/aggregations/bucket/FilterAggregatorTests.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        Directory directory = newDirectory();
+        RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory);
+        int numDocs = randomIntBetween(100, 200);
+        int maxTerm = randomIntBetween(10, 50);
+        int[] expectedBucketCount = new int[maxTerm];
+        Document document = new Document();
+        for (int i = 0; i < numDocs; i++) {
+            if (frequently()) {
+                // make sure we have more than one segment to test the merge
+                indexWriter.commit();
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/bucket/FilterAggregatorTests.java]: you can do `indexWriter.getReader().close()`, which will create a segment too but not do a fsync

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.junit.After;
+import org.junit.Before;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class InternalCardinalityTests extends InternalAggregationTestCase<InternalCardinality> {
+    private static List<HyperLogLogPlusPlus> algos;
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: this variable never seems to be read?

**Review by jpountz — CHANGES_REQUESTED** [review]
I left some minor comments.

File: core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java (hunk 5 above)
Comment (colings86) [inline core/src/test/java/org/elasticsearch/search/aggregations/metrics/cardinality/InternalCardinalityTests.java]: This is needed so we keep a reference to the HyperLogLogPlusPlus implementations we create in `createTestInstance()` so we can close them and release the appropriate `BigArray` pages after the test is complete.

**Review by jpountz — APPROVED** [review]
LGTM

**Pull request merged**